ff/serving/provider/sql.go

```diff
@@ -0,0 +1,1225 @@
+	}
+	columnString := strings.Join(columns, ", ")
+	return store.query.primaryTableCreate(name, columnString), nil
+	//return fmt.Sprintf("CREATE TABLE %s ( %s )", sanitize(name), columnString), nil
+}
+
+func (store *sqlOfflineStore) GetPrimaryTable(id ResourceID) (PrimaryTable, error) {
+	name := GetPrimaryTableName(id)
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, &TableNotFound{id.Name, id.Variant}
+	}
+	columnQuery := store.query.getColumnNames()
+	rows, err := store.db.Query(columnQuery, name)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+	columnNames := make([]TableColumn, 0)
+	for rows.Next() {
+		var column string
+		if err := rows.Scan(&column); err != nil {
+			return nil, err
+		}
+		columnNames = append(columnNames, TableColumn{Name: column})
+	}
+	return &sqlPrimaryTable{
+		db:     store.db,
+		name:   name,
+		schema: TableSchema{Columns: columnNames},
+		query:  store.query,
+	}, nil
+}
+
+func (store *sqlOfflineStore) GetTransformationTable(id ResourceID) (TransformationTable, error) {
+	name := GetTransformationName(id)
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, &TableNotFound{id.Name, id.Variant}
+	}
+	columnQuery := store.query.getColumnNames()
+	rows, err := store.db.Query(columnQuery, name)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+	columnNames := make([]TableColumn, 0)
+	for rows.Next() {
+		var column string
+		if err := rows.Scan(&column); err != nil {
+			return nil, err
+		}
+		columnNames = append(columnNames, TableColumn{Name: column})
+	}
+	return &sqlPrimaryTable{
+		db:     store.db,
+		name:   name,
+		schema: TableSchema{Columns: columnNames},
+		query:  store.query,
+	}, nil
+}
+
+// CreateResourceTable creates a new Resource table.
+// Returns a table if it does not already exist and stores the table ID in the resource index table.
+// Returns an error if the table already exists or if table is the wrong type.
+func (store *sqlOfflineStore) CreateResourceTable(id ResourceID, schema TableSchema) (OfflineTable, error) {
+	if err := id.check(Feature, Label); err != nil {
+		return nil, err
+	}
+
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if exists {
+		return nil, &TableAlreadyExists{id.Name, id.Variant}
+	}
+	tableName := store.getResourceTableName(id)
+	var valueType ValueType
+	if valueIndex := store.getValueIndex(schema.Columns); valueIndex > 0 {
+		valueType = schema.Columns[valueIndex].ValueType
+	} else {
+		valueType = NilType
+	}
+	table, err := store.newsqlOfflineTable(store.db, tableName, valueType)
+	if err != nil {
+		return nil, err
+	}
+	return table, nil
+}
+
+// getValueIndex returns the index of the value column in the schema.
+// Returns -1 if an entity column is not found
+func (store *sqlOfflineStore) getValueIndex(columns []TableColumn) int {
```

Comment:
nit: it's more "go-like" for functions like this to return an (int, error). Probably not worth changing here.